clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -702,9 +702,8 @@ static CompoundStmt *CreateOpenCLKernelBody(Sema &S,
       // initialize them. We create call of __init method and pass built kernel
       // arguments as parameters to the __init method.
       auto getExprForSpecialSYCLObj = [&](const QualType &paramTy,
-                                          FieldDecl *Field,
-                                          const CXXRecordDecl *CRD,
-                                          Expr *Base,
+                                          FieldDecl *Field, Expr *SpecialObjME,
```

Comment:
```suggestion
                                          FieldDecl *Field, MemberExpr *SpecialObjME,
```
The name implies this is going to be a member expression, so we should have it be that.

Reply:
I will move the SpecialObjME parameter to the end with a default value of nullptr.
However, its type needs to remain as Expr*. In the cases where a nullptr is passed in, this function will create a MemberExpr and assign it to SpecialObjME. In the cases where a value is passed in, it is an Expr*, an index operation applied to a MemberExpr. That's why the declared type needs to be the more general Expr*.

Reply:
I don't get what you mean here?  If it is a member expression, it should be stored as one.  It is a pointer, so it shouldn't matter what gets passed in...

Reply:
What's passed in is either nullptr or a member[index]. So what's passed in is not a MemberExpr.
However, when nullptr is passed in, the function creates a MemberExpr to use in place of the nullptr.

Reply:
Then it isn't a Special Object member expression.  The more you've talked about this, the more I'm not sure we have the correct abstraction here.  It seems that passing it like this isn't the correct place to break the ownership model then.

Reply:
I don't know what a Special Object member expression is. I took SpecialObjMe to be a name with nothing special meant by the "Special". I seem to be missing something here.
What change do you suggest?

Reply:
ME typically stands for member expression. I'll have to look during the day tomorrow and better understand this code and the patch before making a suggestion.

Reply:
Also look at the design doc added to sycl/doc, which will explain what is happening here.